docs/spec/spec.md

```diff
@@ -0,0 +1,152 @@
+_Reviewers_: @daneshk  
+_Created_: 2021/11/16  
+_Updated_: 2021/11/16  
+_Issue_: [#2351](https://github.com/ballerina-platform/ballerina-standard-library/issues/2351)
+
+# Introduction
+The UUID library is used to generate UUIDs based on the [RFC 4122](https://www.rfc-editor.org/rfc/rfc4122.html). It is part of Ballerina Standard Library. [Ballerina programming language](https://ballerina.io/) is an open-source programming language for the cloud that makes it easier to use, combine, and create network services.
+
+# Contents
+
+1. [Overview](#1-overview)
+2. [Generating UUIDs](#2-generating-uuids)
+    * 2.1. [Version 1](#21-version-1)
+    * 2.2. [Version 3](#22-version-3)
+    * 2.3. [Version 4](#23-version-4)
+    * 2.4. [Version 5](#24-version-5)
+    * 2.5. [Nil UUID](#25-nil-uuid)
+3. [Checking the Version of UUIDs](#3-checking-the-version-of-uuids)
+4. [Validating UUIDs](#4-validating-uuids)
+5. [UUID Operations](#5-uuid-operations)
+
+## 1. Overview
+This specification elaborates on the functions available in the UUID library.
+
+## 2. Generating UUIDs
+UUID module supports generating 4 versions of UUIDs. They are generated as UUID strings or UUID records. The following is the UUID record definition.
+```
+# Represents a UUID.
+#
+# + timeLow - The low field of the timestamp
+# + timeMid - The middle field of the timestamp
+# + timeHiAndVersion - The high field of the timestamp multiplexed with the version number
+# + clockSeqHiAndReserved - The high field of the clock sequence multiplexed with the variant
+# + clockSeqLo - The low field of the clock sequence
+# + node - The spatially unique node identifier
+public type Uuid readonly & record {
+    ints:Unsigned32 timeLow;
+    ints:Unsigned16 timeMid;
+    ints:Unsigned16 timeHiAndVersion;
+    ints:Unsigned8 clockSeqHiAndReserved;
+    ints:Unsigned8 clockSeqLo;
+    int node;
+};
+```
+
+### 2.1. Version 1
+This is generated using the MAC address of the computer and the time of generation. 
+
+The `uuid:createType1AsString()` function returns a UUID of type 1 as a string.
+```ballerina
+string uuid1 = uuid:createType1AsString();
+```
+
+The `uuid:createType1AsRecord()` function returns a UUID of type 1 as a `Uuid` record.
+```ballerina
+uuid:Uuid uuid1 = check uuid:createType1AsRecord();
+```
+
+### 2.2. Version 3
+This is generated using cryptographic hashing and application-provided text strings. MD5 hashing is used.
+
+UUID library has a set of pre-defined UUIDs strings of well known namespace IDs.
+```
+NAME_SPACE_DNS- Namespace is a fully-qualified domain name
+NAME_SPACE_URL- Namespace is a URL
+NAME_SPACE_OID- Namespace is an ISO OID
+NAME_SPACE_X500- Namespace is an X.500 DN (in DER or a text output format)
+NAME_SPACE_NIL- Empty UUID
+```
+
+The `uuid:createType3AsString()` function returns a UUID of type 3 as a string or else a `uuid:Error`.
+```ballerina
+string uuid3 = check uuid:createType3AsString(uuid:NAME_SPACE_DNS, “ballerina.io”);
```

Comment:
Do we need to assign it to a variable in samples?

Reply:
I think a sample without assigning to the variable doesn't look good.
```
string _ = check uuid:createType3AsString(uuid:NAME_SPACE_DNS, "ballerina.io");
```

Reply:
I was thinking along the line of,
```
check uuid:createType3AsString(uuid:NAME_SPACE_DNS, "ballerina.io");
```

Reply:
Since this is a sample code in spec, better we specify the return type by assigning it to the variable. So it is clear. My suggestion is to do like,
```
string| uuid:Error uuid3 = check uuid:createType3AsString(uuid:NAME_SPACE_DNS, “ballerina.io”);
```

Reply:
even without using the check. because here our intention is to give an idea about our API

Reply:
Fixed in 7ede5a9